### DISCUSSION TITLE
Passing ondemand as argument

### DISCUSSION THREAD

**Discussion by HellsingDarge** [simdjson/simdjson#1790] [Q&A] [answered]
Suppose JSON has the following structure
```json
{ 
  "message_type": "type1", 
  "type1": { 
    "object1": { ... }, "object2": { ... }, "timestamp": 3548, "status": "created" 
  } 
}
```
I need to create an object from `type1`  in JSON, and because the `type1`'s structure is also present in other JSON types I'd like to parse it and construct a class from this data in a method. 
```c++
// doc is ondemand::parser::iterate parsed JSON
auto json = doc["type1"]
MsgType1 obj = fromJson(json)
```
What should the argument's type for `fromJson` be, so that on-demand parsing would work as intended?

**Comment by jkeiser** [comment, marked as answer]
`ondemand::value` (or `simdjson_result<ondemand::value>` if you want to handle any errors yourself).

I suggest making fromJson inline (perhaps even simdjson_really_inline) for performance reasons, because otherwise the compiler will be forced to move some important stuff out of registers and into memory.
